File: blob_detection/blob_detection.py
import cv2
import os
import numpy as np

from optparse import OptionParser

def detect_blob(input_img):
    # Read image file
    input = cv2.imread(input_img)
    cv2.imshow("Input image", input)
    cv2.waitKey(0)
    
    # Create blob detector object
    detector = cv2.SimpleBlobDetector_create()
    # Detect blob
    keypoints = detector.detect(input)
    # Draw blobs as red circle
    blank = np.zeros((1,1))
    blobs = cv2.drawKeypoints(input, keypoints, blank, (0,0,255), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
    # Add description to image
    num_of_blobs = len(keypoints)
    text = "Number of blobs: "+str(num_of_blobs)
    cv2.putText(blobs, text, (20, 550), cv2.FONT_HERSHEY_SIMPLEX, 1, (100, 0, 255), 2)
    cv2.imshow("Blobs using default parameters", blobs)
    cv2.waitKey(0)

    # Create a new blob detector to detect circles
    params = cv2.SimpleBlobDetector_Params()
    
    # Set Area filtering parameters
    params.filterByArea = True
    params.minArea = 100

    # Set Circularity filtering parameters
    params.filterByCircularity = True 
    params.minCircularity = 0.9

    # Set Convexity filtering parameters
    params.filterByConvexity = False
    params.minConvexity = 0.2
        
    # Set inertia filtering parameters
    params.filterByInertia = True
    params.minInertiaRatio = 0.01

    # Create a detector with the parameters
    detector = cv2.SimpleBlobDetector_create(params)
    
    # Detect blob
    keypoints = detector.detect(input)
    # Draw blobs as green circle
    blank = np.zeros((1,1))
    blobs = cv2.drawKeypoints(input, keypoints, blank, (0,255,0), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
    # Add description to image
    num_of_blobs = len(keypoints)
    text = "Number of circles: "+str(num_of_blobs)
    cv2.putText(blobs, text, (20, 550), cv2.FONT_HERSHEY_SIMPLEX, 1, (100, 0, 255), 2)
    cv2.imshow("Blobs using default parameters", blobs)
    cv2.waitKey(0)

if __name__ == '__main__':
    detect_blob("blobs.jpg")

File: blob_detection/src/blob_detection.cpp
#include <iostream>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

void drawBlobs(Mat& input,string& title, vector<KeyPoint>& keyPoints){
	// Draw keypoints
	Mat blobs;
	Scalar redColor(0, 0, 255);
	drawKeypoints(input, keyPoints, blobs, redColor);

	// Add text
	Point bottom(20, 550);
	string text = "Total blobs: " + to_string(keyPoints.size());
	cout << text;
	putText(blobs, text, bottom, FONT_HERSHEY_COMPLEX, 1, redColor, 1);
	imshow("Image with blobs", blobs);
	waitKey(0);

}

void detectBlobs(const string& inputImage){

	// Show input image
	Mat input;
	input = imread(inputImage, CV_LOAD_IMAGE_COLOR);
	imshow("Input image", input);
	waitKey(0);

	// Detect blob
	Ptr<SimpleBlobDetector> blobDetector = SimpleBlobDetector::create();
	vector<KeyPoint> keyPoints;
	blobDetector->detect(input, keyPoints, DrawMatchesFlags::DRAW_RICH_KEYPOINTS);
	string title = "Raw blobs";
	drawBlobs(input, title, keyPoints);


	//Ptr<SimpleBlobDetector::Params> params = SimpleBlobDetector::Params();

	SimpleBlobDetector::Params params;

	params.filterByArea = true;
	params.minArea = 100;

	params.filterByCircularity = true;
	params.minCircularity = 0.9;

	params.filterByConvexity = false;
	params.minConvexity = 0.2;

	params.filterByInertia = true;
	params.minInertiaRatio = 0.01;

	blobDetector = SimpleBlobDetector::create(params);
	vector<KeyPoint> circleKeyPoints;
	blobDetector->detect(input, circleKeyPoints, DrawMatchesFlags::DRAW_RICH_KEYPOINTS);
	string circleTitle = "Circular blobs";
	drawBlobs(input, circleTitle, circleKeyPoints);
}

int main(int argc, char** argv) {
	detectBlobs("../blobs.jpg");
	return 0;
}
